### DISCUSSION TITLE
Stick first line of focused code block to top, for better orientation

### DISCUSSION THREAD

**Discussion by asafm7** [apache/netbeans#8030] [Ideas]
### Description

When the cursor is in a code block, stick the first line of the block to the top when scrolling beyond it for easier orientation. If possible, do the same for nested blocks.

I've seen it in other IDEs. I thought I had also seen it in NetBeans, but I couldn't find anything like it in the settings.

![image](https://github.com/user-attachments/assets/872f341a-2862-4c9d-a7de-90d465bf2784)

### Use case/motivation

I'm looking for an easier understanding of the context within long code blocks.

### Are you willing to submit a pull request?

No

**Comment by Chris2011** [comment]
It is not quite the same in NetBeans, but sometimes, I think when you set the cursor to the last closing bracket. It is not the same, I know but to say we have a similar feature and we can improve this. I really like the idea of it and yes it is very handy. Like it in VS Code.

**Comment by mbien** [comment]
i think this is essentially covered by the classic "breadcrumbs" bar. Maybe an option to move it to the top of the editor would make it easier to use on larger screens.

**Comment by Chris2011** [comment]
![image](https://github.com/user-attachments/assets/e9c76456-50e3-43a3-85e8-9120dd49fa85)
This is more what was looking for, the breadcrumbs are different from this. It shows the whole path from the beginning to the end,

**Reply by mbien** [reply]
took me a while to figure out that this is in fact not a bug. I only noticed this while scrolling and I always thought this is a tooltip getting stuck somewhere. This looks especially weird if a method declaration has multiple lines and you start scrolling after it appeared.

![tooltip](https://github.com/user-attachments/assets/28b993b0-78af-4591-aefe-f22d0c0ffe34)
so much better :)

**Comment by mbien** [comment]
yes I do understand the suggestion. It is semantically equivalent of the breadcrumbs bar, just with reduced functionality.

The main difference is that the breadcrumbs bar is listening to the cursor position. This would render the first line of the clipped block which doesn't fit in the viewport. However, realistically this likely would have to be implemented using a delayed action just like the breadcrumbs bar. So I don't see why it couldn't be used right away for this.

**Comment by asafm7** [comment]
Thanks, @mbien, @Chris2011.

I wasn't aware of the breadcrumb. It is close to what I was looking for, but not quite. Its location isn't optimal. For example, the search bar might buffer between the breadcrumb and the document:

![image](https://github.com/user-attachments/assets/d683eca6-0e9c-4b0e-8bd4-4aa4c9be8d85)

Having the option to move it to the top might bring it closer to what I had in mind.

Also, with [native CSS nesting](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_nesting/Using_CSS_nesting), the breadcrumb only shows the top level:
![image](https://github.com/user-attachments/assets/95473354-15b4-4d2c-883b-ab5480bda2b9)

Is it by design or should I consider it a bug?

**Comment by Chris2011** [comment]
@asafm7 it is just not implemented. When it was implemented, there where no nesting in CSS. You can try scss also whether this is working or not. But it will be a new feautre request, instead of a bug.

**Comment by asafm7** [comment]
Thanks, @Chris2011.

What about the option to change the position of the breadcrumb?

**Comment by Chris2011** [comment]
Sure, but as I said this not fixes the FR request for me either. We have the calculation of when a block goes out of the viewport we just need to stick this to the top. Adding option to move Breadcrumb around is a different FR.

**Comment by asafm7** [comment]
@Chris2011 I agree.

It will be great to have it as a standalone feature. It will be more clear, useful and intuitive.

As you mentioned, when standing on a closing bracket of a top-level block it already happens (the line of the opening bracket sticks to the top) - so the feature is already there, just needs some tweaking.

A couple of notes though:

- The current implementation sticks the line of the opening bracket to the top. Depending on the formatting, this might be useless as this line may contain _only_ the opening bracket.

For example, if CSS is formatted this way:
```
p 
{
  color: red;
  text-align: center;
}
```
The line that sticks to the top is:
```
{
```
- The current implementation only sticks top-level opening brackets to the top. It might be worth considering doing the same with nested blocks.

**Comment by Chris2011** [comment]
I tested your css case in vs code and it also shows just { when you have a new line after the selector. But we don't need to make the same error I justed wanted to have a look how vs code this does. Tomorrow I will also check IntelliJ.

**Comment by asafm7** [comment]
Thanks, @Chris2011.

I assume many platforms didn't consider the case of CSS for this feature, as traditional CSS usually have short blocks. It is relevant though, to SCSS, LESS etc, and it will become even more relevant with the new [native CSS nesting](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_nesting/Using_CSS_nesting).

Let me know if there is anything I can do to help.

**Comment by Chris2011** [comment]
@asafm7 honestly if you are familar with java just have a look into the code. I can try to help to identify the code part where it is. Also your first PR could be to add the option to move the breadcrumb around. What do you think?

**Comment by asafm7** [comment]
Thanks, @Chris2011. Unfortunately, I'm a web developer and not familiar with Java.

BTW, I just noticed that GitHub is doing it nicely, also for nested blocks:

![Screenshot 2024-12-05 200237](https://github.com/user-attachments/assets/6259a693-09d5-4674-8e3f-b73e66de6f14)

**Comment by Chris2011** [comment]
@asafm7 Sure, but did you also check css and also when the bracket is on the new line? I would say let's convert this into a discussion so that we can discuss more, what is not related to this ticket here.

**Comment by asafm7** [comment]
👍 Sounds good

**Comment by Chris2011** [comment]
So at the end, we know that the functionality is still there, it needs enhancement and also an option.

We often see people who likes the new features and others who say: how can I disable it? So yes everything at the end, needs to be configurable. Also this new thing. I guess it is part of the generic editor? Hopefully, if not, then it is part of Java and CSL.
